File: scripts/compare_junctions_hist.R
# compare_junctions_hist.R
# Rscript --vanilla compare_junctions_hist.R <tag>

# load libraries
library(data.table)
library(graphics)
library(plyr)
library(foreach)
library(doParallel)
registerDoParallel(cores=32)

debug = F

system.time({
if (debug){
  tag = paste("_", "default", sep="")
} else {
  # get options tag
  argc = length(commandArgs())
  tag = paste("_", commandArgs(trailingOnly = F)[argc], sep="")

  if ( substr(tag, 2, 3) == "--"){
    stop("Please specify an option tag (e.g. \"default\", \"i20e5\")")
  }
}

## All splicing relevant variants (union of rows from variants.bed files; add column with comma-separated list of sample names)
all_splicing_variants = unique(data.table::fread(paste("all_splicing_variants", tag, ".bed",sep=""), sep = '\t', header = T, stringsAsFactors = FALSE))
colnames(all_splicing_variants) <- c("chrom", "start", "end", "samples")

all_splicing_variants = as.data.table(aggregate(samples ~ chrom + start + end, paste, data=all_splicing_variants))
all_splicing_variants$key <- paste0(all_splicing_variants$chrom, ":", all_splicing_variants$start, "-", all_splicing_variants$end) #this key is just a 1bp-long chrom:start-stop designed to match the regtools output variant_info column

## Get all of the samples
all_samples = strsplit(scan("dir_names.tsv", what="", sep="\n"), "[[:space:]]+")

################################################################################
##### Helper functions #########################################################
################################################################################

# basically combines all the cse_identify_filtered_compare files for this cohort
get_sample_data <- function(sample){
  
  file_name = paste("samples/", sample, "/output/cse_identify_filtered_compare", tag,".tsv", sep = "")
  cse_identify_data = data.table::fread(file_name, sep = '\t', header = T, stringsAsFactors = FALSE, strip.white = TRUE)
  cse_identify_data$sample <- sample
  s <- strsplit(cse_identify_data$variant_info, split = ",")
  cse_identify_data <- data.table::data.table(chrom=rep(cse_identify_data$chrom, sapply(s, length)),
                                              start=rep(cse_identify_data$start, sapply(s, length)),
                                              end=rep(cse_identify_data$end, sapply(s, length)),
                                              name=rep(cse_identify_data$name, sapply(s, length)),
                                              score=rep(cse_identify_data$score, sapply(s, length)),
                                              strand=rep(cse_identify_data$strand, sapply(s, length)),
                                              splice_site=rep(cse_identify_data$splice_site, sapply(s, length)),
                                              acceptors_skipped=rep(cse_identify_data$acceptors_skipped, sapply(s, length)),
                                              exons_skipped=rep(cse_identify_data$exons_skipped, sapply(s, length)),
                                              donors_skipped=rep(cse_identify_data$donors_skipped, sapply(s, length)),
                                              anchor=rep(cse_identify_data$anchor, sapply(s, length)),
                                              known_donor=rep(cse_identify_data$known_donor, sapply(s, length)),
                                              known_acceptor=rep(cse_identify_data$known_acceptor, sapply(s, length)),
                                              known_junction=rep(cse_identify_data$known_junction, sapply(s, length)),
                                              genes=rep(cse_identify_data$genes, sapply(s, length)),
                                              transcripts=rep(cse_identify_data$transcripts, sapply(s, length)),
                                              sample=rep(cse_identify_data$sample, sapply(s, length)),
                                              variant_info=unlist(s))
  return(cse_identify_data)
}

################################################################################
##### Combine cse_identify_filtered_compare.tsv files for each sample ##########
################################################################################

# this is regtools compare output across samples (so it contains variant-junction lines even from samples without variant)
df <- rbindlist(lapply(all_samples, get_sample_data)) # eventually becomes all_cse_identify_data in function a
df$info <- paste(df$chrom, df$start, 
                 df$end, df$anchor, 
                 df$variant_info, sep = "_")

################################################################################
##### Function to create reviewable regtools output files ######################
################################################################################

a <- function(x, all_cse_identify_data){
  ## x comes from all_splicing_variants, so this just gives the samples with the variant
  variant_samples <- unlist(x$samples) 
  
  ##############################################################################
  ##### Get junction and variant information for variant samples ###############
  ##############################################################################
  
  ## Get junctions from concatenated regtools output associated with each variant, only if the data came from a sample actually containing the variant
  variant_junctions_data <- all_cse_identify_data[all_cse_identify_data$variant_info==x$key & all_cse_identify_data$sample %in% variant_samples,]
  
  ## Make dataset with the sample and the junction/variant information
  sample_data <- variant_junctions_data[,c("sample", "info")]
  
  # Normalize junction scores in this variant region for each sample
  summed_variant_scores = variant_junctions_data[, list(score=sum(score)), by=list(sample)]
  colnames(summed_variant_scores) <- c("sample", "score.tmp")
  variant_junctions_data <- merge(variant_junctions_data, summed_variant_scores, by="sample")
  variant_junctions_data$norm_score <- variant_junctions_data$score / variant_junctions_data$score.tmp
  
  # Aggregate data across junction-samples
  if (nrow(variant_junctions_data)[[1]] > 0){ # names column refers to names that regtools gave each junction in the variant output (only will be multiple if variant and junction was found in multiple samples)
    variant_junctions_aggr = variant_junctions_data[, list(names=list(name),mean_norm_score_variant=mean(norm_score),sd_norm_score_variant=sd(norm_score),norm_scores_variant=list(norm_score),total_score_variant=sum(score)), 
                                                    by=list(chrom,start,end,strand,anchor,variant_info,info)]
    
  } else {
    return(data.table())
  }
  
  ## Identify mutated samples
  a <- function(x, tempData){
    return(paste(as.character(tempData[info == x, "sample"]$sample), collapse=",")) # for super-junction aggregated across samples with the variant, get list of samples with variant (and junction)
  }
  variant_junctions_aggr$sample <- sapply(variant_junctions_aggr$info, a, tempData = sample_data)
  
  ##############################################################################
  ##### Get junction and variant information for non-variant samples ###########
  ##############################################################################
  
  ## Samples without variant
  non_samples = setdiff(all_samples, variant_samples)
  non_variant_junctions_data <- all_cse_identify_data[all_cse_identify_data$variant_info==x$key & all_cse_identify_data$sample %in% non_samples]
  
  ## Get variant info
  non_variant_junctions_data$info <- paste(non_variant_junctions_data$chrom, non_variant_junctions_data$start, 
                                           non_variant_junctions_data$end, non_variant_junctions_data$anchor, 
                                           non_variant_junctions_data$variant_info, sep = "_")
  
  ## Make dataset with the sample and the junction/variant information
  non_variant_sample_data <- non_variant_junctions_data[,c("sample", "info")]
  
  # Normalize junction scores in this variant region for each sample
  summed_non_variant_scores = non_variant_junctions_data[, list(score=sum(score)), by=list(sample)]
  colnames(summed_non_variant_scores) <- c("sample", "score.tmp")
  non_variant_junctions_data <- merge(non_variant_junctions_data, summed_non_variant_scores, by="sample")
  non_variant_junctions_data$norm_score <- non_variant_junctions_data$score / non_variant_junctions_data$score.tmp

  # Aggregate data across junction-samples
  if (nrow(variant_junctions_data)[[1]] > 0){
    non_variant_junctions_aggr = non_variant_junctions_data[, list(mean_norm_score_non=mean(norm_score),sd_norm_score_non=sd(norm_score),norm_scores_non=sd(norm_score),total_score_non=sum(score)), 
                                                    by=list(chrom,start,end,strand,anchor,variant_info,info)]
  } else {
    return(data.table())
  }
  
  # Generate histogram of normalized junction scores
  collapsed_norm_scores = non_variant_junctions_data[, list(norm_scores=list(norm_score)), by=list(chrom,start,end,strand)]
  
  # Add non-variant sample aggregate stats to output
  variant_junctions_aggr = merge(x=variant_junctions_aggr, y=non_variant_junctions_aggr,by= 
                c("chrom","start","end","strand","anchor","variant_info","info"),all.x=T)
  
  # split out what p-values can be calculated on vs not
  collapsed_norm_scores$key <- paste(collapsed_norm_scores$chrom, collapsed_norm_scores$start, collapsed_norm_scores$end, collapsed_norm_scores$strand, sep='.')
  variant_junctions_aggr$key <- paste(variant_junctions_aggr$chrom, variant_junctions_aggr$start, variant_junctions_aggr$end, variant_junctions_aggr$strand, sep='.')
  variant_junctions_aggr_run_pvalues <- variant_junctions_aggr[variant_junctions_aggr$key %in% collapsed_norm_scores$key,]
  variant_junctions_aggr_norun_pvalues <- variant_junctions_aggr[!variant_junctions_aggr$key %in% collapsed_norm_scores$key,]
  variant_junctions_aggr_norun_pvalues$pvalue <- 0
  
  calculate_pvalues <- function(junction_data){ #### pull out
    chrom = trimws(junction_data[["chrom"]])
    start = trimws(junction_data[["start"]])
    end = trimws(junction_data[["end"]])
    strand = trimws(junction_data[["strand"]])
    variant_norm_score = junction_data[["mean_norm_score_variant"]]
    ### pass this variable
    i = which(collapsed_norm_scores$chrom == chrom & collapsed_norm_scores$start == start & collapsed_norm_scores$end == end & collapsed_norm_scores$strand == strand)  
    if (length(i)){
      non_variant_norm_scores = collapsed_norm_scores[i]$norm_scores[[1]]
      missing = length(non_samples) - length(non_variant_norm_scores)
      missing_zeros = integer(missing)
      non_variant_norm_scores = c(non_variant_norm_scores, missing_zeros)
      
      variant_norm_score = as.numeric(variant_norm_score)
      all_norm_scores = c(non_variant_norm_scores, variant_norm_score)
      countable = rank(all_norm_scores)
      non_variant_norm_scores_ranked = head(countable, -1)
      variant_norm_score_ranked = tail(countable, 1)
      
      histinfo = hist(non_variant_norm_scores_ranked, 
                      breaks = seq(0.5, max(non_variant_norm_scores_ranked)+1.5, by=1), plot=F)
      mids = histinfo$mids
      cd = cumsum(histinfo$density)
      underestimate = max(which(mids <= variant_norm_score_ranked))
      pvalue = 1-cd[underestimate]
      
      return(pvalue)
    } else {
      return(0)
    }
  }
  if(nrow(variant_junctions_aggr_run_pvalues) > 0){
    # apply above function to each aggregated variant junction
    pvalues = apply(variant_junctions_aggr_run_pvalues, 1, FUN = calculate_pvalues)
    variant_junctions_aggr_run_pvalues$pvalue = pvalues
    
    variant_junctions_aggr <- rbind(variant_junctions_aggr_norun_pvalues, variant_junctions_aggr_run_pvalues)
  } else {
    variant_junctions_aggr <- variant_junctions_aggr_norun_pvalues
  }
  
  return(variant_junctions_aggr) 
}

# for each iteration (i.e. line in all_splicing_variants),
# return a dataframe of regtools info and stats, basically what you end up seeing in regtools_data
# so the samples column comes from all_splicing_variants, since adply uses rbind.fill, which fills in missing columns
regtools_data <- adply(all_splicing_variants, 1, a, all_cse_identify_data=df, .parallel=!debug)

paste_commas <- function(v){
  return(paste(v,collapse = ","))
}

regtools_data <- regtools_data[order(anchor, sample, pvalue)]
regtools_data$norm_scores_variant <- unlist(lapply(regtools_data$norm_scores_variant,paste_commas))
regtools_data$norm_scores_non <- unlist(lapply(regtools_data$norm_scores_non,paste_commas))
regtools_data$samples <- unlist(lapply(regtools_data$samples,paste_commas))
regtools_data$names <- unlist(lapply(regtools_data$names,paste_commas))
regtools_data$sd_norm_score_variant[is.na(regtools_data$sd_norm_score_variant)] = 0
regtools_data$mean_norm_score_non[is.na(regtools_data$mean_norm_score_non)] = 0
regtools_data$sd_norm_score_non[is.na(regtools_data$sd_norm_score_non)] = 0
regtools_data$total_score_variant[is.na(regtools_data$total_score_variant)] = 0
regtools_data$total_score_non[is.na(regtools_data$total_score_non)] = 0

regtools_data = subset(regtools_data, select = !(names(regtools_data) %in% c("key")))
colnames(regtools_data)[colnames(regtools_data)=="sample"] <- "variant_junction_samples"
colnames(regtools_data)[colnames(regtools_data)=="samples"] <- "variant_samples"
colnames(regtools_data)[colnames(regtools_data)=="info"] <- "variant_junction_info"
})

write.table(regtools_data, file=paste("compare_junctions/hist/", "junction_pvalues", tag, ".tsv", sep=""), quote=FALSE, sep='\t', row.names = F)

File: scripts/compare_junctions_hist.py
import csv
from itertools import groupby
import pandas as pd
from dfply import *
import numpy as np
from scipy import stats
import os
import argparse

# input_parser = argparse.ArgumentParser(
#     description="Run RegTools stats script",
# )
# input_parser.add_argument(
#     '-t',
#     '--tag',
#     help="Variant tag parameter used to run RegTools.",
# )
# input_parser.add_argument(
#     '-i',
#     '--variants_file',
#     help="File containing variants to be considered as splicing relevant."
# )
# input_parser.add_argument(
#     '-d',
#     '--dir_names',
#     help="File containing directory names corresponding to each sample that is to be processed."
# )
# input_parser.add_argument(
#     '-v',
#     '--variant-grouping',
#     help="",
#     choices=['strict', 'exclude', 'include']
# )

# args = input_parser.parse_args()

# tag = args.tag
# splicing_variants_inputfile = args.variants_file
# samples_inputfile = args.dir_names
# variant_grouping_mode = args.variant_grouping

tag = 'default'
splicing_variants_inputfile = '/Users/kcotto/Desktop/MET_samples/MET_splicing_variants.bed'
samples_inputfile = '/Users/kcotto/Desktop/MET_samples/samples.txt'
variant_grouping_mode = 'strict'
os.chdir('/Users/kcotto/Desktop/MET_samples/')

# read in all splicing variants
all_splicing_variants = pd.read_csv(
    splicing_variants_inputfile, delimiter='\t', header=None)

# create key to match regtools variant_info column and key2 that is the same as key but with sample name added


def createkey(row):
    key = row[0] + ':' + str(row[1]) + '-' + str(row[2])
    return key

all_splicing_variants['key'] = all_splicing_variants.apply(
    lambda row: createkey(row), axis=1)


def createkey(row):
    key = row[0] + ':' + str(row[1]) + '-' + str(row[2]) + '_' + row[3]
    return key


all_splicing_variants['key2'] = all_splicing_variants.apply(
    lambda row: createkey(row), axis=1)

# read in the sample names
all_samples = []
with open(samples_inputfile, 'r') as samples:
    reader = csv.reader(samples, delimiter='\t')
    for line in reader:
        all_samples.append(line[0])

num_of_samples = len(all_samples)

### read in all of the regtools cse output for this cohort ###
# create list to hold each sample's df
dfs = []

# read each sample's output file into a df and subset columns, split variants into multirows,
# and require that variant is in all_splicing_variants
for sample in all_samples:
    path = f'{sample}/output/cse_identify_filtered_compare_{tag}.tsv'
    df = f'df_{sample}'
    print(f'Reading in {sample}')
    df = pd.read_csv(path, delimiter='\t', header=0)
    df['sample'] = sample
    df = df[['sample', 'variant_info', 'chrom', 'start',
             'end', 'strand', 'anchor', 'score', 'name', 'genes']]
    df = df.dropna(subset=['variant_info'])
    df = df.set_index(['sample', 'chrom', 'start', 'end', 'strand', 'anchor', 'score',
                      'name', 'genes']).apply(lambda x: x.str.split(',').explode()).reset_index()
    df = df.loc[df['variant_info'].isin(all_splicing_variants['key'])]
    dfs.append(df)

# concat all individual dfs into one df
print("Concatenating each sample's df together")
master_df = pd.concat(dfs, axis=0, ignore_index=True)
del dfs

# create various keys


def createkey(row):
    key = row[1] + '_' + str(row[2]) + '_' + \
        str(row[3]) + '_' + row[5] + '_' + row[9]
    return key


master_df['info'] = master_df.apply(lambda row: createkey(row), axis=1)


def createkey(row):
    key = row[9] + '_' + row[0]
    return key


master_df['key'] = master_df.apply(lambda row: createkey(row), axis=1)


def createkey(row):
    key = row[1] + '_' + str(row[2]) + '_' + str(row[3])
    return key


master_df['junction'] = master_df.apply(lambda row: createkey(row), axis=1)

# subset data to work on samples with splicing variant of interest
samples_w_variant_df = master_df.loc[master_df['key'].isin(
    all_splicing_variants['key2'])]
# print(samples_w_variant_df.info(verbose=True))

def add_zeros_variant(row):
    norm_scores = row[1]
    if norm_scores == 0:
        norm_scores = [0]
    samples_wout_variant = row[2]
    samples_w_variant = row[3]    
    num_of_zeros_toadd = num_of_samples - samples_wout_variant - samples_w_variant
    zeros = np.repeat(0, num_of_zeros_toadd).tolist()
    norm_scores = norm_scores + zeros
    norm_scores.sort(reverse=True)
    new_norm_score_value = (',').join(map(str, norm_scores))
    return new_norm_score_value

# tmp_df['new_norm_scores'] = tmp_df.apply(lambda row: add_zeros_nonvariant(row), axis=1)

# start performing the calculations for this subset of data
print('Calculating normalized scores for samples with variants of interest')
if variant_grouping_mode == 'include':
    samples_w_variant_df = (samples_w_variant_df >>
                            group_by(X.key) >>
                            summarize(score_tmp=X.score.sum()) >>
                            outer_join(samples_w_variant_df, by='key')
                            )
    samples_w_variant_df['norm_score'] = samples_w_variant_df['score'] / \
        samples_w_variant_df['score_tmp']
    samples_w_variant_df = (samples_w_variant_df >>
                            group_by('junction') >>
                            summarize(mean_norm_score_variant=X.norm_score.mean(), sd_norm_score_variant=X.norm_score.std(), total_score_variant=X.score.sum()) >>
                            outer_join(samples_w_variant_df, by='junction')
                            )
    tmp_df = samples_w_variant_df.groupby('junction')[
        ['norm_score', 'score', 'variant_info', 'sample', 'name']].aggregate(lambda x: x.tolist()).reset_index()
    samples_w_variant_df = pd.merge(
        samples_w_variant_df, tmp_df, on='junction')
    samples_w_variant_df = samples_w_variant_df[['sample_y', 'variant_info_y', 'chrom', 'start', 'end', 'strand', 'anchor',
                                                 'info', 'genes', 'name_y', 'mean_norm_score_variant', 'sd_norm_score_variant',
                                                 'norm_score_y', 'score_y', 'junction', 'total_score_variant']]
    samples_w_variant_df.columns = ['junction_samples', 'variant_info', 'chrom', 'start', 'end', 'strand', 'anchor',
                                    'info', 'genes', 'names', 'mean_norm_score_variant', 'sd_norm_score_variant',
                                    'norm_scores_variant', 'scores', 'junction', 'total_score_variant']
    samples_w_variant_df['variant_samples'] = samples_w_variant_df['junction_samples']
    samples_w_variant_df = samples_w_variant_df[~samples_w_variant_df.astype(
        str).duplicated()]
else:
    samples_w_variant_df = (samples_w_variant_df >>
                            group_by(X.key) >>
                            summarize(score_tmp=X.score.sum()) >>
                            outer_join(samples_w_variant_df, by='key')
                            )
    samples_w_variant_df['norm_score'] = samples_w_variant_df['score'] / \
        samples_w_variant_df['score_tmp']
    samples_w_variant_df = (samples_w_variant_df >>
                            group_by('info') >>
                            summarize(mean_norm_score_variant=X.norm_score.mean(), sd_norm_score_variant=X.norm_score.std(), total_score_variant=X.score.sum()) >>
                            outer_join(samples_w_variant_df, by='info')
                            )
    tmp_df = samples_w_variant_df.groupby('info')[['norm_score', 'score', 'sd_norm_score_variant',
                                                   'mean_norm_score_variant', 'sample', 'name']].aggregate(lambda x: x.tolist()).reset_index()
    samples_w_variant_df = pd.merge(samples_w_variant_df, tmp_df, on='info')
    samples_w_variant_df = samples_w_variant_df[['sample_x', 'sample_y', 'variant_info', 'chrom', 'start', 'end', 'strand', 'anchor',
                                                 'info', 'genes', 'name_y', 'mean_norm_score_variant_y', 'sd_norm_score_variant_y',
                                                 'norm_score_y', 'score_y', 'junction', 'total_score_variant']]
    samples_w_variant_df.columns = ['sample_x', 'sample_y', 'variant_info', 'chrom', 'start', 'end', 'strand', 'anchor',
                                    'info', 'genes', 'names', 'mean_norm_score_variant', 'sd_norm_score_variant',
                                    'norm_scores_variant', 'scores', 'junction', 'total_score_variant']
    tmp_df = samples_w_variant_df.groupby('variant_info')[['sample_x']].aggregate(lambda x: list(set(x.tolist()))).reset_index()
    samples_w_variant_df = pd.merge(samples_w_variant_df, tmp_df, on='variant_info')
    samples_w_variant_df = samples_w_variant_df[['sample_y', 'variant_info', 'chrom', 'start', 'end', 'strand', 'anchor',
                                                 'info', 'genes', 'names', 'mean_norm_score_variant', 'sd_norm_score_variant',
                                                 'norm_scores_variant', 'scores', 'junction', 'total_score_variant', 'sample_x_y']]
    samples_w_variant_df.columns = ['junction_samples', 'variant_info', 'chrom', 'start', 'end', 'strand', 'anchor',
                                    'info', 'genes', 'names', 'mean_norm_score_variant', 'sd_norm_score_variant',
                                    'norm_scores_variant', 'scores', 'junction', 'total_score_variant', 'variant_samples']
    samples_w_variant_df = samples_w_variant_df[~samples_w_variant_df.astype(
        str).duplicated()]

# work on samples that don't have the variant of interest
print('Calculating normalized scores for samples without variants of interest')
samples_wout_variant_df = master_df[~master_df['key'].isin(
    all_splicing_variants['key2'])]
del (master_df)

# mode = 'strict' #others include 'include' and 'exclude'
# if mode == 'strict':
samples_wout_variant_df = (samples_wout_variant_df >>
                           group_by(X.key) >>
                           summarize(score_tmp=X.score.sum()) >>
                           outer_join(samples_wout_variant_df, by='key')
                           )
samples_wout_variant_df['norm_score'] = samples_wout_variant_df['score'] / \
    samples_wout_variant_df['score_tmp']
samples_wout_variant_df = samples_wout_variant_df.loc[samples_wout_variant_df['variant_info'].isin(
    all_splicing_variants['key'])]
tmp_df = samples_wout_variant_df.groupby(
    'info')[['norm_score', 'sample']].aggregate(lambda x: x.tolist()).reset_index()
samples_wout_variant_df = pd.merge(samples_wout_variant_df, tmp_df, on='info')
samples_wout_variant_df['samples_wout_variant_count'] = samples_wout_variant_df['norm_score_y'].astype(
    str).str.count(',') + 1
if variant_grouping_mode == 'include' or variant_grouping_mode == 'exclude':
    samples_wout_variant_df = samples_wout_variant_df[~samples_wout_variant_df['junction'].isin(
        samples_w_variant_df['junction'])]
    tmp_df = samples_wout_variant_df.groupby(
        'info')[['norm_score_x']].aggregate(lambda x: x.tolist()).reset_index()
    samples_wout_variant_df = pd.merge(
        samples_wout_variant_df, tmp_df, on='info')
    samples_wout_variant_df = (samples_wout_variant_df >>
                           group_by('info') >>
                           summarize(total_score_non=X.score.sum()) >>
                           outer_join(samples_wout_variant_df, by='info')
                           )
    print(samples_wout_variant_df.info())
    samples_wout_variant_df = samples_wout_variant_df[['sample_y', 'variant_info', 'chrom', 'start', 'end', 'strand', 'anchor',
                                                       'info', 'genes', 'norm_score_x_y', 'junction', 'total_score_non', 'samples_wout_variant_count']]
else:
    samples_wout_variant_df = (samples_wout_variant_df >>
                           group_by('info') >>
                           summarize(total_score_non=X.score.sum()) >>
                           outer_join(samples_wout_variant_df, by='info')
                           )
    samples_wout_variant_df = samples_wout_variant_df[['sample_y', 'variant_info', 'chrom', 'start', 'end', 'strand', 'anchor',
                                                       'info', 'genes', 'norm_score_y', 'junction', 'total_score_non', 'samples_wout_variant_count']]
samples_wout_variant_df.columns = ['sample', 'variant_info', 'chrom', 'start', 'end', 'strand', 'anchor',
                                'info', 'genes', 'norm_scores_non', 'junction', 'total_score_non', 'samples_wout_variant_count']

print('Merging dataframes')
master_df = pd.merge(samples_w_variant_df, samples_wout_variant_df, how='left' ,on='info')
master_df = master_df[-master_df.astype(
        str).duplicated()]
del(samples_wout_variant_df)
del(samples_w_variant_df)

master_df['samples_w_variant_count'] = master_df['variant_samples'].astype(
    str).str.count(',') + 1

tmp_df = master_df[['info', 'norm_scores_non', 'samples_wout_variant_count', 'samples_w_variant_count']]
tmp_df = tmp_df.fillna(0)

def add_zeros_nonvariant(row):
    norm_scores = row[1]
    if norm_scores == 0:
        norm_scores = [0]
    samples_wout_variant = row[2]
    samples_w_variant = row[3]    
    num_of_zeros_toadd = num_of_samples - samples_wout_variant - samples_w_variant
    zeros = np.repeat(0, num_of_zeros_toadd).tolist()
    norm_scores = norm_scores + zeros
    norm_scores.sort(reverse=True)
    new_norm_score_value = (',').join(map(str, norm_scores))
    return new_norm_score_value

tmp_df['new_norm_scores'] = tmp_df.apply(lambda row: add_zeros_nonvariant(row), axis=1)
master_df = pd.merge(master_df, tmp_df, how='left' ,on='info')
del(tmp_df)

def get_mean(row):
    values = row[33].split(',')
    values = [float(i) for i in values]
    mean = np.mean(values)
    return mean

master_df['mean_norm_score_non'] = master_df.apply(lambda row: get_mean(row), axis=1)

def get_sd(row):
    values = row[33].split(',')
    values = [float(i) for i in values]
    std = np.std(values)
    return std

master_df['sd_norm_score_non'] = master_df.apply(lambda row: get_sd(row), axis=1)

print('getting p-values for associations')

def get_min(row):
    values = row[12]
    values = [float(i) for i in values]
    minimum = min(values)
    return(minimum)

master_df['min_norm_score_variant'] = master_df.apply(lambda row: get_min(row), axis=1)

def get_pvalue_mean(row):
    values = row[33].split(',')
    values = [float(i) for i in values]
    mean_value = row[10]
    pvalue = stats.percentileofscore(values, mean_value)
    pvalue = 1 - (pvalue/100.0)
    pvalue = re.sub('[\[\]]', '', np.array_str(pvalue))
    return pvalue

master_df['p_value_mean'] = master_df.apply(lambda row: get_pvalue_mean(row), axis=1)

def get_pvalue_min(row):
    values = row[33].split(',')
    values = [float(i) for i in values]
    min_value = row[36]
    pvalue = stats.percentileofscore(values, min_value)
    pvalue = 1 - (pvalue/100.0)
    pvalue = re.sub('[\[\]]', '', np.array_str(pvalue))
    return pvalue
    
master_df['p_value_min'] = master_df.apply(lambda row: get_pvalue_mean(row), axis=1)

master_df = master_df[['variant_samples', 'variant_info_x', 'genes_x', 'junction_samples', 
                       'chrom_x', 'start_x', 'end_x', 'strand_x', 'anchor_x', 'info', 
                       'names', 'mean_norm_score_variant', 'sd_norm_score_variant', 
                       'norm_scores_variant', 'total_score_variant', 'mean_norm_score_non',
                       'sd_norm_score_non', 'new_norm_scores', 'total_score_non', 'p_value_mean','p_value_min']]
master_df.columns = ['variant_samples', 'variant_info', 'genes', 'junction_samples', 
                       'chrom', 'start', 'end', 'strand', 'anchor', 'variant_junction_info', 
                       'names', 'mean_norm_score_variant', 'sd_norm_score_variant', 
                       'norm_scores_variant', 'total_score_variant', 'mean_norm_score_non',
                       'sd_norm_score_non', 'norm_scores_non', 'total_score_non', 'p_value_mean','p_value_min']

master_df = master_df.applymap(lambda x: x[0] if isinstance(x, list) else x)
master_df = master_df.fillna(0)

master_df.to_csv(f'junction_pvalues_{tag}_{variant_grouping_mode}.tsv', sep='\t', index=False)
print(master_df.info())